crates/cheatnet/src/execution/cheatable_syscall_handler.rs

```diff
@@ -0,0 +1,297 @@
+};
+use cairo_vm::hint_processor::hint_processor_definition::HintProcessorLogic;
+use cairo_vm::vm::runners::cairo_runner::ResourceTracker;
+use cairo_vm::{
+    hint_processor::hint_processor_definition::HintReference,
+    serde::deserialize_program::ApTracking,
+    types::{exec_scope::ExecutionScopes, relocatable::Relocatable},
+    vm::{
+        errors::{hint_errors::HintError, vm_errors::VirtualMachineError},
+        runners::cairo_runner::RunResources,
+        vm_core::VirtualMachine,
+    },
+};
+use starknet_api::hash::StarkFelt;
+use std::{any::Any, collections::HashMap};
+
+pub type SyscallSelector = DeprecatedSyscallSelector;
+
+// This hint processor modifies the standard syscalls implementation to react upon changes
+// introduced by cheatcodes that e.g. returns mocked data
+// If it cannot execute a cheatcode it falls back to SyscallHintProcessor, which provides standard implementation of
+// hints from blockifier
+pub struct CheatableSyscallHandler<'a> {
+    pub syscall_handler: SyscallHintProcessor<'a>,
+    pub cheatnet_state: &'a mut CheatnetState,
+}
+
+impl<'a> CheatableSyscallHandler<'a> {
+    pub fn new(
+        syscall_handler: SyscallHintProcessor<'a>,
+        cheatnet_state: &'a mut CheatnetState,
+    ) -> Self {
+        CheatableSyscallHandler {
+            syscall_handler,
+            cheatnet_state,
+        }
+    }
+}
+
+// crates/blockifier/src/execution/syscalls/hint_processor.rs:472 (ResourceTracker for SyscallHintProcessor)
+impl ResourceTracker for CheatableSyscallHandler<'_> {
+    fn consumed(&self) -> bool {
+        self.syscall_handler.context.vm_run_resources.consumed()
+    }
+
+    fn consume_step(&mut self) {
+        self.syscall_handler.context.vm_run_resources.consume_step();
+    }
+
+    fn get_n_steps(&self) -> Option<usize> {
+        self.syscall_handler.context.vm_run_resources.get_n_steps()
+    }
+
+    fn run_resources(&self) -> &RunResources {
+        self.syscall_handler
+            .context
+            .vm_run_resources
+            .run_resources()
+    }
+}
+
+impl HintProcessorLogic for CheatableSyscallHandler<'_> {
+    fn execute_hint(
+        &mut self,
+        vm: &mut VirtualMachine,
+        exec_scopes: &mut ExecutionScopes,
+        hint_data: &Box<dyn Any>,
+        constants: &HashMap<String, Felt252>,
+    ) -> HintExecutionResult {
+        let maybe_extended_hint = hint_data.downcast_ref::<Hint>();
+
+        if let Some(Hint::Starknet(StarknetHint::SystemCall { .. })) = maybe_extended_hint {
+            if let Some(Hint::Starknet(hint)) = maybe_extended_hint {
+                return self.execute_next_syscall_cheated(vm, hint);
+            }
+        }
+
+        match contract_print(vm, maybe_extended_hint) {
+            PrintingResult::Printed => return Ok(()),
+            PrintingResult::Passed => (),
+            PrintingResult::Err(err) => return Err(err),
+        }
```

Comment:
We probably should implement a separate syscall handler for this to be consistent with `TestExecutionSyscallHandler`. Also we can consider exposing other forge_std functions in contracts if it would be useful.

Reply:
Can be left for future work.